### DISCUSSION TITLE
Tips for avoiding the API rate limit

### DISCUSSION THREAD

**Discussion by skedwards88** [community/community#77255] [Discover: GitHub Best Practices]
## About rate limits

GitHub uses rate limits to ensure that the API remains available for all users.

There are two types of rate limits: primary and secondary. Primary rate limits determine the total number of requests that you can make in one hour. Secondary rate limits determine how quickly you can make multiple requests within a period of time. Visit the documentation to learn more about rate limits for the [REST API](https://docs.github.com/rest/overview/rate-limits-for-the-rest-api) and [GraphQL API](https://docs.github.com/graphql/overview/rate-limits-and-node-limits-for-the-graphql-api).

## Tips to avoid exceeding the rate limits

- Don't poll the API. Instead, use [webhooks](https://docs.github.com/webhooks) to get notified when events occur.
- Make requests serially instead of concurrently.
- If you are making a large number of requests that modify data on GitHub, wait at least one second between each request.
- If you are requesting data that might not have changed since the last time that you requested it, use [conditional requests](https://docs.github.com//rest/guides/best-practices-for-using-the-rest-api#use-conditional-requests-if-appropriate).

## What to do if you exceed the rate limit

If you receive a rate limit error, you should stop making requests temporarily according to these guidelines:

- If the `retry-after` response header is present, you should not retry your request until after that many seconds has elapsed.
- If the `x-ratelimit-remaining` header is `0`, you should not make another request until after the time specified by the `x-ratelimit-reset` header. The `x-ratelimit-reset` header is in UTC epoch seconds.
- Otherwise, wait for at least one minute before retrying. If your request continues to fail due to a secondary rate limit, wait for an exponentially increasing amount of time between retries, and throw an error after a specific number of retries.

## Get a higher rate limit

You can't increase your secondary rate limit, but you can increase your primary rate limit:

- Make authenticated requests instead of unauthenticated requests. Authenticated requests have a significantly higher rate limit than unauthenticated requests.
- Use a [GitHub App](https://docs.github.com/apps/creating-github-apps/about-creating-github-apps/about-creating-github-apps) if your integration is serving an organization with many users or repositories. The rate limit for GitHub Apps using an installation access token scales with the number of repositories and number of organization users.
- Upgrade to [GitHub Enterprise Cloud](https://docs.github.com/enterprise-cloud@latest/admin/overview/about-github-enterprise-cloud) if you are using GitHub Apps or OAuth apps. GitHub Apps or OAuth apps have higher rate limits for organizations that use GitHub Enterprise Cloud.

**Comment by jmikrut** [comment]
Hi @skedwards88 - have there been any recent changes with GitHub rate limiting? My product creates repositories on behalf of our users, using a GitHub app, and authenticating on behalf of the user.

In the last hour, we have started being rate limited across the board as we try and create a repository on behalf of our user. Our platform is now unable to create new repos. Would love to know if something changed.

**Reply by skedwards88** [reply]
No recent changes to rate limiting, just sharing some tips here. The rate limit for GitHub App user access token is combined with any requests that another GitHub App or OAuth app makes on that user's behalf and any requests that the user makes with a personal access token, so high usage from other tokens is a potential cause. If these [best practice](https://docs.github.com/en/rest/guides/best-practices-for-using-the-rest-api) and [troubleshooting](https://docs.github.com/en/rest/overview/troubleshooting-the-rest-api) tips don't help, GitHub Support may be able to help.

**Reply by 07510480632** [reply]
J

**Comment by 07510480632** [comment]
> j

**Comment by harshai07** [comment]


**Reply by ghostinhershell** [reply]
Hi @harshai07, 

We’ve clarified our stance on using generative AI tools like ChatGPT within our Community via [this announcement](https://github.com/orgs/community/discussions/69006). Please review the guidelines to ensure your post meets them as failure to adhere to those rules can result in action taken by our moderator team.  You can read our updated [Code of Conduct](https://github.com/community/community/blob/baa5cd9020ef3f2bbf86dca1191d9c90f9b0822d/CODE_OF_CONDUCT.md#reasonable-use-of-ai-generated-content) and the [announcement](https://github.com/orgs/community/discussions/69006) for more details. Thank you for helping us maintain an authentic and beneficial space for everyone.

**Comment by AndrewOfC** [comment]
Quick way of checking your rate limit for the REST API:

https://github.com/AndrewOfC/githubpy#quick-rate-limit-check

**Comment by alsheroziy** [comment]
Hello everyone, I just joined the github community, please tell me the rules

**Comment by devstein** [comment]
Hi @skedwards88 👋 thank you for the resources! 

> > Upgrade to [GitHub Enterprise Cloud](https://docs.github.com/enterprise-cloud@latest/admin/overview/about-github-enterprise-cloud) if you are using GitHub Apps or OAuth apps. GitHub Apps or OAuth apps have higher rate limits for organizations that use GitHub Enterprise Cloud.

Does upgrading to GitHub Enterprise Cloud only impact the rate limits of GitHub apps _we are using_? Or does it also impact the rate limits for _our Github app that our customers use_

**Comment by 4roblox** [comment]
How do I remove my rate limit on roblox?
